### PR TITLE
SALTO-4581 Configuration for Broken Outgoing References From Fetched Instances

### PR DESCRIPTION
**Pull request by tomsellek**
Currently if we fetch an instance and it has a reference that we can't resolve we drop the whole instance. This PR makes the behavior configurable, such that what we do when encountering an invalid reference is determined by the type of the reference target.
This is supported by new config values as described in the Jira ticket.

---

Note the 'heuristic' we use to determine the type of the reference target - if there's a single entry in the field's `referenceTo` annotation, we use that entry. Otherwise we try to guess the type from the internal ID.

---
_Release Notes_: 
Salesforce: Introduced the 'brokenOutgoingReferencesSettings' data configuration section. Use it to control what happens when Salto encounters an invalid reference in a data record.

---
_User Notifications_: 
Salesforce: the default behavior when encountering invalid references in data records changed. Before, we would drop instances with invalid references. Now we retain them. If you want to retain the old behavior, set the 'brokenOutgoingReferencesSettings.defaultBehavior' data configuration value to "ExcludeInstance".

### CODE REVIEW

File: packages/salesforce-adapter/src/filters/custom_object_instances_references.ts [hunk 1]
```diff
@@ -183,47 +197,55 @@ const replaceLookupsWithRefsAndCreateRefMap = async (
...
+                type: await apiName(await instance.getType(), true),
+                id: await apiName(instance),
+                field,
+              },
+              targetId: instance.value[field.name],
+            })
+            return value
+          }
+          default: {
+            throw new Error('Unrecognized broken refs behavior. Is the configuration valid?')
```
Comment (adamro) [inline packages/salesforce-adapter/src/filters/custom_object_instances_references.ts:244]: We should be able to validate and avoid this

File: packages/salesforce-adapter/config_doc.md [hunk 2]
```diff
@@ -207,7 +240,7 @@ salesforce {
 |---------------------------------------------------------------|-------------------------------------------------------|----------------------------------------------------------------------------------------------------|
 | [polling](#client-polling-options)                            | `{}` (no overrides)                                   | Configuration for polling asynchronous operations (deploy, retrieve, bulk data operations)         |
 | [deploy](#client-deploy-options)                              | `{}` (no overrides)                                   | Deploy options                                                                                     |
-| [retry](#retry-configuration-options)                         | `{}` (no overrides)                                   | Configuration for retrying on errors                                                               |
+| [retry](#client-retry-options)                                | `{}` (no overrides)                                   | Configuration for retrying on errors                                                               |
```
[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/config_doc.md:243]: It's not, just boyscouting. The link was broken before.

File: packages/salesforce-adapter/src/filters/custom_object_instances_references.ts [hunk 3]
```diff
@@ -272,6 +294,20 @@ const getIllegalRefSources = (
   return illegalRefSources
 }
 
+const buildCustomObjectPrefixKeyMap = async (
```
[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/src/filters/custom_object_instances_references.ts:297]: Kinda? It's our fallback if the `referenceTo` annotation has more than one entry.
Anyway, this same code used to be in this very file until I removed it in SALTO-4500 (it was used for creating the error messages), so presumably it works to some extent. I can try and see if I can test it with some edge cases.

File: packages/salesforce-adapter/src/filters/custom_object_instances_references.ts [hunk 4]
```diff
@@ -272,6 +294,20 @@ const getIllegalRefSources = (
...
+  elements: Element[]
+): Promise<Record<string, string>> => {
+  const objectTypesWithKeyPrefix = elements
+    .filter(isObjectType)
+    .filter(objectType => isDefined(objectType.annotations[KEY_PREFIX]))
```
[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/src/filters/custom_object_instances_references.ts:302]: Yes, if an instance of the type that's missing a key prefix is referenced from a field with more than one entry in its `referenceTo` annotation. Do you think I should a warning if that happens?

File: packages/salesforce-adapter/config_doc.md
[the remark this answers is not in the document]
Comment (adamro) [reply packages/salesforce-adapter/config_doc.md:1]: Whatever we want *new* CPQ users to have in their config
cc: @tomermevorach

File: packages/salesforce-adapter/src/filters/custom_object_instances_references.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamro) [reply packages/salesforce-adapter/src/filters/custom_object_instances_references.ts:302]: Yea, probably worth a warning

File: packages/salesforce-adapter/src/filters/custom_object_instances_references.ts (hunk 1 above)
Comment (adamro) [reply packages/salesforce-adapter/src/filters/custom_object_instances_references.ts:244]: If it's typed and you covered all the options? 🤔

File: packages/salesforce-adapter/src/types.ts [hunk 5]
```diff
@@ -218,14 +229,42 @@ const saltoAliasSettingsType = new ObjectType({
...
+})
+
+const brokenOutgoingReferencesSettingsType = new ObjectType({
+  elemID: new ElemID(constants.SALESFORCE, 'brokenOutgoingReferencesSettings'),
+  fields: {
+    defaultBehavior: {
+      refType: BuiltinTypes.STRING,
+    },
+    perTargetOverrides: {
+      refType: new MapType(BuiltinTypes.STRING),
```
[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/src/types.ts:251]: image

File: packages/salesforce-adapter/src/filters/custom_object_instances_references.ts (hunk 1 above)
Comment (tomsellek) [reply packages/salesforce-adapter/src/filters/custom_object_instances_references.ts:244]: Yeah, the linter still complains.
When I do
image

I get
image

Comment (adamro) [reply packages/salesforce-adapter/src/filters/custom_object_instances_references.ts:244]: I am actually not sure why or how but it doesn't *really* matter :)

File: packages/salesforce-adapter/src/fetch_profile/data_management.ts
Comment (adamro) [inline packages/salesforce-adapter/src/fetch_profile/data_management.ts:1]: Do we have anything to add to `validateDataManagementConfig` with the latest configs btw?

**Review by adamro — APPROVED** [review]
One last question about validation + the documentation and default config stuff

But otherwise LGTM so approving to save a cycle

File: packages/salesforce-adapter/src/fetch_profile/data_management.ts
Comment (tomsellek) [reply packages/salesforce-adapter/src/fetch_profile/data_management.ts:1]: Hmm...
I think the NaCl-level restrictions are enough.

File: packages/salesforce-adapter/src/filters/custom_object_instances_references.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/src/filters/custom_object_instances_references.ts:302]: OK, I added a warning but only in the case where the missing type is actually used anywhere.

File: packages/salesforce-adapter/src/filters/custom_object_instances_references.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/src/filters/custom_object_instances_references.ts:311]: I added a debug log. 
Until we change the defaults, the only potential issue is if there's a field that references both `User` and something else (which we've seen happen - e.g. `User` + `Group`) and we don't have these types' key prefixes - currently, we do `InternalId` for such fields, but after this PR it would be `ExcludeInstance`.

[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/src/filters/custom_object_instances_references.ts:311]: In my environment (not doing partial fetches of any kind) I get `226/973 object types have a key prefix` and the `User` and `Group` types both have a key prefix. So this is a non-issue on at least one environment :)

File: packages/salesforce-adapter/config_doc.md
[the remark this answers is not in the document]
Comment (tomermevorach) [reply packages/salesforce-adapter/config_doc.md:1]: @adamro @tomsellek any reason not to use the same defaults?

[the remark this answers is not in the document]
Comment (adamro) [reply packages/salesforce-adapter/config_doc.md:1]: @tomermevorach What do you want the "new" default for new users will be? "InternalId" or "BrokenReference"?

[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/config_doc.md:1]: @tomermevorach ?

[the remark this answers is not in the document]
Comment (tomermevorach) [reply packages/salesforce-adapter/config_doc.md:1]: I think it should be internal id like today. Its also indicated in the ticket: https://salto-io.atlassian.net/browse/SALTO-4581

Any reason to change it? Arent users replicated on refresh and therefore preserve their internal ids?

[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/config_doc.md:1]: @tomermevorach isn't the default today `ExcludeInstance`? I think that's what the discussion in the ticket is about?

[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/config_doc.md:1]: @adamro I discussed this with @tomermevorach and the bottom line is that this default (`BrokenReference` by default, `InternalId` for `User`) is correct for CPQ.
To start with we'll make the default `ExcludeInstance` just to keep the behavior the same, and we'll change it later along with customer communication.

[the remark this answers is not in the document]
Comment (tomsellek) [reply packages/salesforce-adapter/config_doc.md:1]: Following further discussions between @adamro and @tomermevorach , we'll make the default in `config_creator.ts` be `BrokenReference` so that new users will get the benefit of the new behavior.
